pkg/operator/resources/utils/common.go

```diff
@@ -95,6 +97,7 @@ func CreateDeployment(name, matchKey, matchValue, serviceAccountName string, ima
 		ImagePullSecrets: imagePullSecrets,
 	}
 	deployment := ResourceBuilder.CreateDeployment(name, "", matchKey, matchValue, serviceAccountName, replicas, podSpec, infraNodePlacement)
```

Comment:
So I think this is a bug here, infraNodePlacement is shared with the CDI CR object, so when we add the anti affinity it happens on the same object over and over again, causing extra affinities to show up. I would personally modify this to look something like this:
```go
func CreateDeployment(name, matchKey, matchValue, serviceAccountName string, imagePullSecrets []corev1.LocalObjectReference, replicas int32, infraNodePlacement *sdkapi.NodePlacement) *appsv1.Deployment {
	podSpec := corev1.PodSpec{
		SecurityContext: &corev1.PodSecurityContext{
			RunAsNonRoot: &[]bool{true}[0],
		},
		ImagePullSecrets: imagePullSecrets,
	}
	inpCopy := infraNodePlacement.DeepCopy()
	inpCopy.Affinity = AddPodPreferredDuringSchedulingIgnoredDuringExecution(name, infraNodePlacement.Affinity)
	deployment := ResourceBuilder.CreateDeployment(name, "", matchKey, matchValue, serviceAccountName, replicas, podSpec, inpCopy)
	return deployment
}
```

And modify `AddPodPreferredDuringSchedulingIgnoredDuringExecution` to return an affinity and take an affinity so we pass that to the function that creates the deployment. This way the deployment gets the proper anti affinity that you are looking for. Something like this:
```go
// AddPodPreferredDuringSchedulingIgnoredDuringExecution to affinity
func AddPodPreferredDuringSchedulingIgnoredDuringExecution(name string, affinity *corev1.Affinity) *corev1.Affinity {
	var affinityCopy *corev1.Affinity
	preferredDuringSchedulingIgnoredDuringExecution := corev1.WeightedPodAffinityTerm{
		Weight: int32(1),
		PodAffinityTerm: corev1.PodAffinityTerm{
			LabelSelector: &metav1.LabelSelector{
				MatchExpressions: []metav1.LabelSelectorRequirement{
					{
						Key:      "cdi.kubevirt.io",
						Operator: metav1.LabelSelectorOpIn,
						Values:   []string{name}},
				},
			},
			TopologyKey: "kubernetes.io/hostname",
		},
	}

	if affinity != nil && affinity.PodAntiAffinity != nil {
		affinityCopy = affinity.DeepCopy()
		affinityCopy.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution = append(affinityCopy.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution, preferredDuringSchedulingIgnoredDuringExecution)
	} else if affinity != nil {
		affinityCopy = affinity.DeepCopy()
		affinityCopy.PodAntiAffinity = &corev1.PodAntiAffinity{
			PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{preferredDuringSchedulingIgnoredDuringExecution},
		}
	} else {
		affinityCopy = &corev1.Affinity{
			PodAntiAffinity: &corev1.PodAntiAffinity{
				PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{preferredDuringSchedulingIgnoredDuringExecution},
			},
		}
	}
	return affinityCopy
}
```
Now we are working with proper copies of objects and this passes for when I test.

Reply:
Thank you very much for your answer. So this is a very hidden problem caused by pointers.This gave me great inspiration.

Reply:
IMO it is a downright bug, we just never hit it because we didn't set any pod affinity rules. I think the anti affinity is important when doing multiple replicas, so I fully support adding those. It just uncovered that issue.